### ISSUE TITLE
Some combo of a linkerscript, minimal sections copied, and a .eh_frame section cause r2 to see file as only 0xff bytes

### ISSUE THREAD

**Issue by anall** [radare/radare2#9420]
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Ubuntu x86 64
| File format of the file you reverse (mandatory)      | ELF
| Architecture/bits of the file (mandatory)            | x86/32
| r2 -v full output, **not truncated** (mandatory)         | radare2 2.4.0-git 17303 @ linux-x86-64 git.2.2.0-494-g047d67289<br>commit: 047d672898817d10f93e24948602ccfe91a2f830 build: 2018-02-17__22:23:10
| r2 -V full output in a pastebin document (mandatory) | https://gist.github.com/anall/09df186c680d540a3e8bb6e4e18a5580
| gcc -v | gcc version 7.2.0 (Ubuntu 7.2.0-8ubuntu3)
| ld -v | GNU ld (GNU Binutils for Ubuntu) 2.29.1

### Expected behavior

`r2` should actually show the contents of the binary.

### Actual behavior

`r2` shows the entire file as `ff`

### Steps to reproduce the behavior 

[minimal.zip](https://github.com/radare/radare2/files/1734412/minimal.zip)

The attached `test.elf` should exit immediately and with a return code of 100. I'm honestly surprised that Linux even sneezes at this executable, let alone lets it have a single CPU cycle -- but it does.

This is a minimized test case -- I have minimized this as far as I can and still reproduce the error.
As far as I can tell, this is some combination of a linker script, C code, and a `.eh_frame` section.

The `.eh_frame` section seems to be the true cause here.

Minimal linker.ld:

    ENTRY(somewhere)
    SECTIONS {
      . = 1M;
      .text : {
        *(.text*)
      }
      . = ALIGN(4K);
      .bss : { *(COMMON) }
    }

Minimal `foo.c` file:

    void somewhere() {
      asm( "mov $1,%eax\n"
        "mov $100,%ebx\n"
        "int $0x80" );
      return;
    }

Build with:

    g++ -m32 -c foo.c
    ld -o test.elf -melf_i386 -nostdlib -T linker.ld linker.ld foo.o

If you remove the `.eh_frame` section from `foo.o` before linking, you get a `.elf` that r2 likes:

    objcopy --remove-section .eh_frame foo.o foo.fixed.o
    ld -o test_good.elf -melf_i386 -nostdlib -T linker.ld linker.ld foo.fixed.o

This file has been attached as `test_good.elf`.

You cannot remove the `.eh_frame` section *after* linking -- doing so will still result in an ELF that exhibits the issue ( see `test_postlink.elf` )

I apologize, but this is as far as I can get this reduced and still trigger the issue.

Thank you for your time.

### Additional Logs, screenshots, source-code,  configuration dump, ...

![screenshot_2018-02-17_22-20-56](https://user-images.githubusercontent.com/12371/36348243-e00f305c-1430-11e8-8945-64235e3766a2.png)

**Comment by leberus** [comment]
I'm taking a look at this. I made some progress.
I'll post soon.

**Comment by leberus** [comment]
I think that this hack within <b>init_phdr()</b> is wrong (not the hack itself but its usage here)

<pre>
static int init_phdr(ELFOBJ *bin) {
...
        bool linux_kern_hack = false;
        /* Enable this hack only for the X86 64bit ELFs */
        const int _128K = 1024 * 128;
        if (bin->b->length > _128K && bin->ehdr.e_machine == EM_X86_64 ||
                bin->ehdr.e_machine == EM_386) {
                linux_kern_hack = true;
        }       
...
...
        /* Here is the where all the fun starts.
         * Linux kernel since 2005 calculates phdr offset wrongly
         * adding it to the load address (va of the LOAD0).
         * See `fs/binfmt_elf.c` file this line:
         *    NEW_AUX_ENT(AT_PHDR, load_addr + exec->e_phoff);
         * So after the first read, we fix the address and read it again
         */
        if (linux_kern_hack) {
                ut64 load_addr = Elf_(r_bin_elf_get_baddr) (bin);
                bin->ehdr.e_phoff = Elf_(r_bin_elf_v2p) (bin, load_addr + bin->ehdr.e_phoff);
                linux_kern_hack = false;
                goto read_phdr;
        }
</pre>

I don't really get why do we have this hack here.
<b>AT_PHDR</b> is being set to the value <pre>vaddr + phdr_offset</pre> so the linker (or whatever)
can look for the program header inside of the memory image of the program once the program has been loaded into memory.
Maybe I'm missing something, but this doesn't make sense here.
In <b>init_phdr()</b> we're reading from a buffer we have previously filled in Elf_(r_bin_elf_new), not from the program's memory.

This works most of the times because in <pre>Elf_(r_bin_elf_v2p)</pre>, after the computation, we end up
returning the same phdr offset.

But with the current example, the computation returns a wrong phdr offset, so we end up reading from a wrong offset within the binary img, and this affects for instance <pre>Elf_(r_bin_elf_get_entry_offset)</pre>.

With the hack removed:

<pre>
osalvador@d104:~/lab/r2/9420> r2 ./test.elf -c 'pd'
Warning: Cannot initialize dynamic strings
            ;-- entry0:
            ;-- section..text:
            ;-- section.LOAD0:
            ;-- section.ehdr:
            ;-- somewhere:
            0x00100000      55             push ebp                    ; [10] m-rw- section size 52 named ehdr
            0x00100001      89e5           mov ebp, esp
            0x00100003      e814000000     call sym.__x86.get_pc_thunk.ax
            0x00100008      0564000000     add eax, 0x64               ; 'd'
            0x0010000d      b801000000     mov eax, 1
            0x00100012      bb64000000     mov ebx, 0x64               ; 'd' ; 100
            0x00100017      cd80           int 0x80
            0x00100019      90             nop
            0x0010001a      5d             pop ebp
            0x0010001b      c3             ret
</pre>

As I said, I might be missing something, but why do we have this hack here if we're not reading from the program's memory but from a buffer which contains the binary?
@radare

**Comment by Maijin** [comment]
@leberus could you check if this works too with this patch removed https://github.com/radare/radare2-regressions/blob/master/t.formats/elf/elf-phdrlinuxtrick and make a PR?

**Comment by leberus** [comment]
Removing the hack did make elf-phdrlinuxtrick fail.
That case is indeed quite tricky. I found the paper [here](https://gist.github.com/pzread/2ae0bb3aa5fe0dc69fcf3257c41db944).
I came up with:
<pre>
diff --git a/libr/bin/format/elf/elf.c b/libr/bin/format/elf/elf.c
index 00ae05e91..8c83f79af 100644
--- a/libr/bin/format/elf/elf.c
+++ b/libr/bin/format/elf/elf.c
@@ -161,6 +161,7 @@ static int init_phdr(ELFOBJ *bin) {
                return false;
        }
 
+       bool phdr_found = false;
        bool linux_kern_hack = false;
        /* Enable this hack only for the X86 64bit ELFs */
        const int _128K = 1024 * 128;
@@ -178,6 +179,9 @@ read_phdr:
                        return false;
                }
                bin->phdr[i].p_type = READ32 (phdr, j)
+               if (bin->phdr[i].p_type == PT_PHDR) {
+                       phdr_found = true;
+               }
 #if R_BIN_ELF64
                bin->phdr[i].p_flags = READ32 (phdr, j)
                bin->phdr[i].p_offset = READ64 (phdr, j)
@@ -203,7 +207,7 @@ read_phdr:
         *    NEW_AUX_ENT(AT_PHDR, load_addr + exec->e_phoff);
         * So after the first read, we fix the address and read it again
         */
-       if (linux_kern_hack) {
+       if (linux_kern_hack && phdr_found) {
                ut64 load_addr = Elf_(r_bin_elf_get_baddr) (bin);
                bin->ehdr.e_phoff = Elf_(r_bin_elf_v2p) (bin, load_addr + bin->ehdr.e_phoff);
                linux_kern_hack = false;
</pre>

Basically, it there isn't a program <pre>PT_PHDR</pre>, we stick to the one defined in <pre>elf_header->e_phoff</pre>
This doesn't break https://github.com/radare/radare2-regressions/blob/master/t.formats/elf/elf-phdrlinuxtrick and let r2 show the binary of this bug correctly.
Anyway, I'd like to hear from @XVilka and see what he thinks about that before moving forward (he might have a better idea).

**Comment by Maijin** [comment]
Could you make a PR ?

**Comment by XVilka** [comment]
yes, looks good to me

**Comment by anall** [comment]
I can confirm that fixes both my reduced testcase, as well as the original I reduced from.

**Comment by leberus** [comment]
I think we can close this one, right?
